### DISCUSSION TITLE
apiParser.unpkgHost问题咨询

### DISCUSSION THREAD

**Discussion by leegsen7** [umijs/dumi#1676] [Q&A] [answered]
> 其中 `unpkgHost` 配置项用于自定义 `unpkg.com` 的地址以加快访问速度，比如自己的私有镜像地址。解析过程中如果存在找不到的依赖，会兜底到 `unpkgHost` 的地址去找。

我自己本地删除了node_modules，执行yarn安装后，断网是能够build成功的。但在构建服务器上build是一直卡住的（该服务器禁止访问外网，有自建npm服务，但无类unpkg服务）。

想了解一下，`存在找不到的依赖` 这种是什么情况，又有哪些包是需要去兜底找的

`dumi-afx-deps` 这个包无源码真看不懂里面的实现😂

**Comment by PeachScript** [comment, marked as answer]
> 想了解一下，`存在找不到的依赖` 这种是什么情况，又有哪些包是需要去兜底找的

正好今天在看这部分实现，通常来说只要是不隔离幽灵依赖的包管理器，比如（`npm`/`yarn`），并且为没有提供类型的依赖正确安装了 `@types/xxx` 的包，那么大部分情况下解析器是能找到依赖，不去 unpkg 找的；如果有一些极端情况仍然需要去找，例如三方包导出的明明依赖 `@types/abc` 但却没有声明它

> `dumi-afx-deps` 这个包无源码真看不懂里面的实现😂

这个包是压缩输出的，因为解析器相关团队暂时没有做开源的计划和相应人力投入，所以可以理解为黑盒不可调试的版本，社区同学也没法参与贡献😞

**Reply by leegsen7** [reply]
明白，我再捣鼓捣鼓，感谢回复。🍉

**Comment by leegsen7** [comment]
终于是捣鼓出来了。

猜测 `dumi-afx-deps` 会对 `unpkgHost` 地址预先进行访问检查，而我们pipeline机器是禁止访问外网的，导致一致卡住。

实际上我们用yarn安装，理论上是不用 `unpkgHost` 兜底，这里随便填个内网地址就行了。

神奇~
